src/testing/TestFields.tsx

```diff
@@ -0,0 +1,151 @@
+import React, { FC } from 'react';
+import { Box, Text, Label, Input, Checkbox, Textarea, Radio } from 'theme-ui';
+import { FormikContextType, FormikValues } from 'formik';
+import { Message } from 'retranslate';
+import * as yup from 'yup';
+
+import { TestType, ObjectSchema, FieldSchema } from '../api';
+
+interface TestFieldsProps {
+  form: FormikContextType<FormikValues>;
+  testType: TestType;
+}
+
+export const TestFields: FC<TestFieldsProps> = ({ form, testType }) => {
+  const fieldError = (key: keyof TestType['resultsSchema']['properties']) =>
+    form.touched[key] && form.errors[key] ? <Text>{form.errors[key]}</Text> : null;
+
+  return (
+    <>
+      {Object.entries(testType.resultsSchema.properties).map(([key, value]) => {
+        if (value.oneOf) {
+          return (
+            <Box key={key}>
+              {value.title} {testType.resultsSchema.required?.includes(key) && '*'}
+              {value.oneOf.map((option) => (
+                <Label key={`${option.const}`}>
+                  <Radio
+                    checked={form.getFieldProps(key).value === option.const}
+                    value={typeof option.const === 'number' ? option.const : `${option.const}`}
+                    onChange={({ target }) =>
+                      form.setFieldValue(
+                        key,
+                        typeof option.const === 'boolean' ? target.value === 'true' : target.value
+                      )
+                    }
+                  />
+                  {option.title}
+                </Label>
+              ))}
+              {fieldError(key)}
+            </Box>
+          );
+        }
+
+        if (value.type === 'boolean') {
```

Comment:
Yeah, by having type: "boolean" in your oneOf schema, I think you're going to render a checkbox and a radio control?

Reply:
Oh, actually no, because you return...

Reply:
Recommendations here are very welcome, as it was a bit tricky to get to work.

Reply:
The way I have this stuff implemented is something more like this (pseudocode):
```
useState(model, setModel, props.model);

const onPropertyChange = (property, value) => {
  const newModel = { ...model, [property] = value};
  setModel(newModel);
  this.onChange(newModel);
});

props.schema.properties.map((childSchema, property) => {
  <Field 
    schema={childSchema} 
    model={model[property]}
    onChange={(value) => onPropertyChange(property, value) 
  />
});
```
Where the Field component handles the control type switch, and any data type mapping issues (like radio needing a string).

Reply:
In the real one we also pass an isValid with the on Change

`this.onChange(newModel, isValid)`

Which can be calculated with an off the shelf JSON schema validator.  That makes it easy to do stuff like enabling/disabling submit button.

Reply:
Added a `TODO` to extract general components out.